frontends/p4/duplicateHierarchicalNameCheck.cpp

```diff
@@ -0,0 +1,81 @@
+/*
+Copyright 2024 Cisco Systems, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "duplicateHierarchicalNameCheck.h"
+
+#include "lib/log.h"
+
+namespace P4 {
+
+cstring DuplicateHierarchicalNameCheck::getName(const IR::IDeclaration *decl) {
+    return decl->getName();
+}
+
+const IR::Node *DuplicateHierarchicalNameCheck::postorder(IR::Annotation *annotation) {
+    if (annotation->name != IR::Annotation::nameAnnotation) return annotation;
+
+    cstring name = annotation->getName();
+    if (!name.startsWith(".")) {
+        if (!stack.empty()) {
+            name = absl::StrCat(absl::StrJoin(stack, ".",
+                                              [](std::string *out, cstring s) {
+                                                  absl::StrAppend(out, s.string_view());
+                                              }),
+                                ".", name.string_view());
+        }
+    }
+    // The node the annotation belongs to
+    CHECK_NULL(getContext()->parent);
+    auto *annotatedNode = getContext()->parent->node;
+    CHECK_NULL(annotatedNode);
+    // variable and struct declarations are fine if they have identical
+    // name annotations, and such name annotations can be synthesized by
+    // p4c before this pass.  Ignore them.
+    if (annotatedNode->is<IR::Declaration_Variable>() || annotatedNode->is<IR::Type_Struct>()) {
+        return annotation;
+    }
+    bool foundDuplicate = false;
+    auto *otherNode = annotatedNode;
+    if (annotatedNode->is<IR::P4Table>()) {
+        if (annotatedTables.count(name)) {
```

Comment:
You could probably save one map lookup here, e.g.:
```c++
auto [it, inserted] = annotatedTables.insert(name, annotatedNode);
if (!inserted) {
  foundDuplicate = true;
  otherNode = *it;
}
```

Same for code below.

Probable the code could be deduplicated as well? Something like this:
```c++
auto checkForDuplicates = [&foundDuplicate, &otherNode](auto &namesToCheck, const IR::Node *annotatedNode) {
  auto [it, inserted] = namesToCheck.insert(name, annotatedNode);
  if (!inserted) {
    foundDuplicate = true;
    otherNode = *it;
  }
}
```

And then just call this lambda in each of this `is` checks:
```c++
if (annotatedNode->is<IR::P4Table>() {
  checkForDuplicates(annotatedTables, annotatedNode);
} else () ...
```